### PR TITLE
Dropdown component (feature branch)

### PR DESCRIPTION
**Pull request by MelSumner**
## :pushpin: Summary

If merged, this PR adds the dropdown component.

***

## :white_check_mark: Reviewer's checklist

- [ ] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1201758147765505
  - https://app.asana.com/0/0/1201880481980935

### CODE REVIEW

File: packages/components/addon/components/hds/dropdown/list-item.hbs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    {{#if @copyItemTitle}}
+      <div class="hds-dropdown-list-item--copy-item-title">{{@copyItemTitle}}</div>
+    {{/if}}
+    <button
+      type="button"
+      class="{{if @state (concat 'is-' @state)}} foo {{if this.isSuccess 'is-success'}}"
+      {{on "click" this.copyCode}}
+    >
+      <div class="hds-dropdown-list-item__copy-item-text">
+        {{@text}}
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.hbs:22]: good catch!

File: .changeset/gorgeous-singers-punch.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"@hashicorp/design-system-components": minor
+---
+
+added a stacking context for the “Button” component so that the focus z-index is isolated in the button
```
[the remark this answers is not in the document]
Comment (Dhaulagiri) [reply .changeset/gorgeous-singers-punch.md:5]: no idea how it got here 🤔  probably a stray from the merging from main process. 

what we could do is repurpose this for the changeset for this PR which we could do by editing the content of this markdown file with the changelog message we want for this component.

File: packages/components/addon/components/hds/dropdown/list-item.hbs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          <div class="hds-dropdown-list-item__interactive-icon">
+            <FlightIcon @name={{@icon}} @isInlineBlock={{false}} />
+          </div>
+        {{/if}}
+        <div class="hds-dropdown-list-item__interactive-text">
+          {{@text}}
+        </div>
+      </a>
+    {{else}}
+      <button class="{{if @state (concat 'is-' @state)}}" type="button" ...attributes>
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.hbs:72]: we have to pass ...attributes so they can add the {{on "click" this.something}} but we don't need it in the others. WDYT?

File: packages/components/addon/components/hds/dropdown/list-item.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+export const DEFAULT_COLOR = 'action';
+export const DEFAULT_ITEM = 'interactive';
+export const COLORS = ['action', 'critical'];
+export const ITEMS = [
+  'copy-item',
+  'description',
+  'generic',
+  'interactive',
+  'separator',
+  'title',
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.js:15]: the order is alphabetically sorted and it is the same everywhere. I am not sure where the disconnect is on this, can you say more?

File: packages/components/addon/components/hds/dropdown/list-item.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    if (this.args.item === 'generic') {
+      // eslint-disable-next-line getter-return
+      return;
+    } else {
+      assert(
+        '@text for "Hds::Dropdown::ListItem" must have a valid value',
+        text !== undefined
+      );
+    }
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.js:37]: it's not throwing the assertion at all, need to debug more.

File: packages/components/addon/components/hds/dropdown/list-item.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    return item;
+  }
+
+  /**
+   * Get the class names to apply to the component.
+   * @method classNames
+   * @return {string} The "class" attribute to apply to the component.
+   */
+  get classNames() {
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.js:90]: the action is always the last thing in the component backing class. the classNames can come right before that, though.

File: .changeset/gorgeous-singers-punch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply .changeset/gorgeous-singers-punch.md:5]: should I remove the file? should the changeset be automatically generated?

File: packages/components/addon/components/hds/dropdown/list-item.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    return classes.join(' ');
+  }
+
+  @action
+  copyCode() {
+    navigator.clipboard.writeText(this.args.text);
+    // this if statement resolves to [object Promise] so maybe some improvements
+    // could be made here
+    if (navigator.clipboard.readText()) {
+      this.isSuccess = true;
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.js:111]: I had some console.log stuff in here but took it out; everything is functioning as expected for right now. I am going to wait until we get some post-integration test feedback from product teams before making any other changes.

File: packages/components/addon/components/hds/dropdown/list-item.hbs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  </li>
+
+{{else if (eq this.item "generic")}}
+  <li class={{this.classNames}}>
+    {{yield}}
+  </li>
+
+{{else if (eq this.item "copy-item")}}
+  <li class={{this.classNames}}>
+    {{#if @copyItemTitle}}
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.hbs:13]: since `@title` already exists, I thought that would be a bad idea. The next best thing that came to mind was to add title to the thing that it is, since that felt a little intuitive (tried not to overthink it). Maybe this is something we should ask devs for feedback on? Whether or not the name makes sense in context? WDYT?

File: packages/components/addon/components/hds/dropdown/index.hbs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    <ul
+      class="hds-dropdown-list
+        {{if (eq @listPosition 'left') 'hds-dropdown-list--position-left' 'hds-dropdown-list--position-right'}}"
+    >
+      {{yield
+        (hash
+          CopyItem=(component "hds/dropdown/list-item" item="copy-item")
+          Description=(component "hds/dropdown/list-item" item="description")
+          Generic=(component "hds/dropdown/list-item" item="generic")
+          ListItem=(component "hds/dropdown/list-item")
```
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: hmmm you make a point there. I wonder if we lose the whole _idea_ of list-item (it was intended to convey an li element) if we change this? Does it even matter anymore? Do we want dev feedback? WDYT?

File: .changeset/gorgeous-singers-punch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dhaulagiri) [reply .changeset/gorgeous-singers-punch.md:5]: Happy to pair on this as well, but this step is the manual part of the process so my suggestion would be fairly similar to what would happen if you deleted the file and re-ran `yarn changeset`. Not a terrible idea if you wanted to do that though as it'd give you a chance to try out that workflow.

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 9 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: they're alllllll sorted alphabetically. If we want to change it here, we should change it for all the components and update this in the docs (unless that was already removed)

[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: if we don't want to lose the idea of item we have a few options:
- we prefix the entities with `ListItem::` (so would be `ListItem::Interactive
- we yield the list-items to a `<:List>` named block and we use it in the invocation of the component ([see my thread here](https://hashicorp.slack.com/archives/C11JCBJTW/p1649413297164169))

[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: In the `hbs` file have a different order, for example: `packages/components/addon/components/hds/dropdown/list-item.hbs`.

Also, in Figma they're not sorted alphabetically, but by context/usage.

File: packages/components/addon/components/hds/dropdown/index.hbs [hunk 10]
```diff
@@ -0,0 +1,38 @@
+<Hds::Disclosure class="hds-dropdown">
+  <:toggle as |t|>
+    {{#if (eq @toggle "toggleButton")}}
```
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: For context and future reference: [thread here](https://hashicorp.slack.com/archives/C11JCBJTW/p1649413297164169).

File: packages/components/addon/components/hds/dropdown/list-item.hbs (hunk 8 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.hbs:13]: OK, let's decide once we decide if split or not the `list-item` component in single sub-components.

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 9 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: I think both of those add additional complexity. I guess if these are our only options then we should change it to interactive.

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: let's go ahead and change it. We can let teams try it out and if it feels confusing we can switch it back. WDYT?

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 9 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:31]: I don't have to maintain the Figma file :) 
I think it's enough to have it match Figma on API.

File: packages/components/app/styles/components/dropdown.scss [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+  &:focus,
+  &.is-focus {
+    @include hds-focus-ring-with-pseudo-element(
+      $top: -1px,
+      $right: -1px,
+      $bottom: -1px,
+      $left: -1px,
+      $radius: $hds-dropdown-toggle-border-radius
+    );
+  }
```
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:64]: It works, visually, but look at the generated CSS 
screenshot_1289

File: packages/components/app/styles/components/dropdown.scss [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+  align-items: center;
+  border-radius: 3px; // $hds-dropdown-toggle-border-radius- 1px padding - 1px border
+  display: flex;
+  justify-content: center;
+  height: 2rem;
+  padding: 1px;
+  width: 2rem;
+
+  img {
+    border-radius: inherit;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:83]: if we add an `object-fit: cover` we are resilient to the possibility that the image may not be square.

```suggestion
    border-radius: inherit;
    object-fit: cover;
```

File: packages/components/app/styles/components/dropdown.scss [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+    border-color: var(--token-color-border-strong);
+  }
+}
+
+.hds-dropdown-toggle-icon__wrapper {
+  align-items: center;
+  border-radius: 3px; // $hds-dropdown-toggle-border-radius- 1px padding - 1px border
+  display: flex;
+  justify-content: center;
+  height: 2rem;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:78]: [issue] you have partially converted the sizes to `rem`s but that was not the agreement we reached time ago: for now we still use `px` apart from `font-size` (rem) and `line-height` (unitless)

File: packages/components/app/styles/components/dropdown.scss [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+
+.hds-dropdown-list {
+  background-color: var(--token-color-surface-primary);
+  border-radius: 6px;
+  box-shadow: var(--token-surface-high-box-shadow);
+  list-style: none;
+  margin: 0;
+  max-width: 25rem;
+  min-width: 12.5rem;
+  padding: 0.25rem 0;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:114]: [issue] rems instead of px

File: packages/components/app/styles/components/dropdown.scss [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+  &.hds-dropdown-list--position-right {
+    position: absolute;
+    right: 0;
+    top: calc(100% + 0.25rem);
+    z-index: 2; // https://github.com/hashicorp/design-system/issues/114
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:119]: [issue] this in theory should have a proper value, 2 is just a "random" number here. maybe look into Structure or Cloud UI and see what value of Z they are using for this, and let's use the same

File: packages/components/app/styles/components/dropdown.scss [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+
+// LIST > LIST-ITEM
+// HDS::DROPDOWN::LIST-ITEM
+
+.hds-dropdown-list-item--title {
+  color: var(--token-color-foreground-strong);
+  font-family: var(--token-typography-body-100-font-family);
+  font-size: var(--token-typography-body-100-font-size);
+  font-weight: var(--token-typography-font-weight-semibold);
+  line-height: var(--token-typography-body-100-line-height);
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:138]: [suggestion] not for now, but when there are standard sizes/styles for the text like in this case, is better to use the predefined CSS helper classes (eg. in this case it would be `hds-typography-body-100 hds-font-weight-semibold`

File: packages/components/app/styles/components/dropdown.scss [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+  padding-left: 1rem;
+  padding-right: 1rem;
+}
+
+.hds-dropdown-list-item--description {
+  color: var(--token-color-foreground-faint);
+  font-family: var(--token-typography-body-100-font-family);
+  font-size: var(--token-typography-body-100-font-size);
+  font-weight: var(--token-typography-font-weight-regular);
+  line-height: var(--token-typography-body-100-line-height);
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:152]: see comment above about standard styles

File: packages/components/app/styles/components/dropdown.scss [hunk 18]
```diff
@@ -0,0 +1,410 @@
...
+  color: var(--token-color-foreground-faint);
+  font-family: var(--token-typography-body-100-font-family);
+  font-size: var(--token-typography-body-100-font-size); // 13
+  font-weight: var(--token-typography-font-weight-semibold);
+  line-height: var(--token-typography-body-100-line-height); // 18
+  padding: 0.125rem 0 0.25rem;
+}
+
+.hds-dropdown-list-item--copy-item {
+  box-sizing: border-box;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:181]: it can be removed now
```suggestion
```

File: packages/components/app/styles/components/dropdown.scss [hunk 19]
```diff
@@ -0,0 +1,410 @@
...
+  padding: 0.625rem 1rem 0.75rem;
+  width: 100%;
+
+  button {
+    @include hds-focus-ring-basic();
+    
+    background-color: transparent;
+    border-radius: 5px;
+    border: 1px solid var(--token-color-border-primary);
+    box-sizing: border-box;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:191]: it can be removed now


```suggestion
```

File: packages/components/app/styles/components/dropdown.scss [hunk 20]
```diff
@@ -0,0 +1,410 @@
...
+
+.hds-dropdown-list-item--interactive {
+  position: relative;
+  isolation: isolate; // used to create a new stacking context (needed to have the pseudo element below text/icon but not the parent container)
+  min-height: 36px;
+
+  // need to reset a few extra things to make the button visually appear the same as the links
+  // TODO this is 0.125rem taller than the link...
+  button {
+    box-sizing: border-box;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:252]: it can be removed now

```suggestion
```

File: packages/components/app/styles/components/dropdown.scss [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+  padding: 0.125rem 0 0.25rem;
+}
+
+.hds-dropdown-list-item--copy-item {
+  box-sizing: border-box;
+  padding: 0.625rem 1rem 0.75rem;
+  width: 100%;
+
+  button {
+    @include hds-focus-ring-basic();
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:186]: [chore/consistency] move this declaration just before the other states

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: ok after reading more comments and seeing the thing about nesting yet another named block in here, I think we should leave this exactly like it is. It's quickly spilling over into over-engineered territory.

File: packages/components/app/styles/components/dropdown.scss [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+      background-color: var(--token-color-surface-interactive-hover);
+      cursor: pointer;
+    }
+
+    &:focus,
+    &.is-focus {
+      //TODO this focus is just way too complex 
+      background-color:  var(--token-color-surface-action);
+      border-color: var(--token-color-focus-action-internal);
+    }
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:210]: there is an issue with this focus, because is adding a double visual border. I'll have a look next week and see to fix it

File: packages/components/app/styles/components/dropdown.scss [hunk 23]
```diff
@@ -0,0 +1,410 @@
...
+    &.is-success {
+      border-color: var(--token-color-border-success);
+      background-color: var(--token-color-surface-success);
+      
+      .hds-dropdown-list-item__copy-item-icon {
+        color: var(--token-color-foreground-success);
+      }
+    } 
+
+    .hds-dropdown-list-item__copy-item-text {
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:226]: [chore] this can be un-indented (no need to live under the `.hds-dropdown-list-item--copy-item` block

File: packages/components/app/styles/components/dropdown.scss [hunk 24]
```diff
@@ -0,0 +1,410 @@
...
+      font-size: var(--token-typography-code-100-font-size);
+      font-weight: var(--token-typography-font-weight-regular);
+      line-height: var(--token-typography-code-100-line-height);
+      // max-width: 250px; // TODO we should be able to figure out the proportions here
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:230]:
```suggestion
```

File: packages/components/app/styles/components/dropdown.scss [hunk 25]
```diff
@@ -0,0 +1,410 @@
...
+      overflow: hidden;
+      text-align: left;
+      text-overflow: ellipsis;
+    }
+
+    .hds-dropdown-list-item__copy-item-icon {
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:236]: [chore] this can be un-indented (no need to live under the `.hds-dropdown-list-item--copy-item` block

File: packages/components/app/styles/components/dropdown.scss [hunk 26]
```diff
@@ -0,0 +1,410 @@
...
+    } 
+
+    .hds-dropdown-list-item__copy-item-text {
+      font-size: var(--token-typography-code-100-font-size);
+      font-weight: var(--token-typography-font-weight-regular);
+      line-height: var(--token-typography-code-100-line-height);
+      // max-width: 250px; // TODO we should be able to figure out the proportions here
+      overflow: hidden;
+      text-align: left;
+      text-overflow: ellipsis;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:233]: still missing the `white-space: nowrap;` declaration, otherwise the text will be multiline

```suggestion
      text-overflow: ellipsis;
      white-space: nowrap;
```

(notice: if we want to support also multiline, we should expose a prop to control this behaviour)

File: packages/components/app/styles/components/dropdown.scss [hunk 27]
```diff
@@ -0,0 +1,410 @@
...
+  }
+}
+
+.hds-dropdown-list-item--interactive {
+  position: relative;
+  isolation: isolate; // used to create a new stacking context (needed to have the pseudo element below text/icon but not the parent container)
+  min-height: 36px;
+
+  // need to reset a few extra things to make the button visually appear the same as the links
+  // TODO this is 0.125rem taller than the link...
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:250]: I'll try to fix it next week

File: packages/components/app/styles/components/dropdown.scss [hunk 28]
```diff
@@ -0,0 +1,410 @@
...
+  button {
+    box-sizing: border-box;
+    border: 1px inset transparent; // cause of the extra height
+    background-color: transparent;
+    width: 100%;
+    &:hover {
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:256]:
```suggestion

    &:hover {
```

File: packages/components/app/styles/components/dropdown.scss [hunk 29]
```diff
@@ -0,0 +1,410 @@
...
+      }
+    }
+  }
+}
+
+.hds-dropdown-list-item__interactive-text {
+  font-family: var(--token-typography-body-200-font-family);
+  font-size: var(--token-typography-body-200-font-size);
+  font-weight: var(--token-typography-font-weight-medium);
+  line-height: var(--token-typography-body-200-line-height);
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:336]: we can use CSS helper classes for these

File: packages/components/app/styles/components/dropdown.scss [hunk 30]
```diff
@@ -0,0 +1,410 @@
...
+    }
+
+    &:hover,
+    &.is-hover {
+      color: var(--token-color-foreground-action-hover);
+      &::before {
+        background-color: currentColor;
+      }
+    }
+    &:active,
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:363]:
```suggestion

    &:active,
```

**Review by didoo — COMMENTED** [review]
@MelSumner this is the second part of the code review. no major issues, everything can be fixed. 

Only thing I ask you to update are the sizes in `rem`. In this file you have updated some of them and now part of them are `rem`s, others are `px`s. There was a "soft" agreement time ago that we would use `px` apart from font-size (`rem`) and line-height (unitless)

context: https://hashicorp.slack.com/archives/C11JCBJTW/p1642168594144300?thread_ts=1641913069.030800&cid=C11JCBJTW

screenshot_1290

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: I don't agree. Everyone that commented in the thread is saying the current implementation is not ergonomic. Why would you go with that solution?

File: packages/components/addon/components/hds/dropdown/list-item.hbs (hunk 3 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.hbs:72]: I think that we have to document this (for the users), because is not obvious at all and they don't have a way to know how this works unless they read the documentation or look at the source code.

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.js:95]: If you look at the codebase, in all the components that have both `@action` declarations and `classNames` method, the `classNames` method is the last.

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.js:117]: this code as is now is not correct. as you mentioned above, `writeText` returns a promise so probably the check on the read should be asynchronous (and it should check we read what we expected)

I'll try to work on this next week

File: packages/components/addon/components/hds/dropdown/index.hbs [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+    <ul
+      class="hds-dropdown-list
+        {{if (eq @listPosition 'left') 'hds-dropdown-list--position-left' 'hds-dropdown-list--position-right'}}"
+    >
+      {{yield
+        (hash
+          CopyItem=(component "hds/dropdown/list-item" item="copy-item")
+          Description=(component "hds/dropdown/list-item" item="description")
+          Generic=(component "hds/dropdown/list-item" item="generic")
+          Interactive=(component "hds/dropdown/list-item")
```
Comment (didoo) [inline packages/components/addon/components/hds/dropdown/index.hbs:31]: You could do this and remove the default from the backing class, more clear (and semantically correct, now that we are yielding the components there's no more "default").
```suggestion
          Interactive=(component "hds/dropdown/list-item" item="interactive")
```

**Review by didoo — COMMENTED** [review]
added a few more comments

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/list-item.js:95]: it's incorrect then. tracked properties > everything else > actions, in that order. we can say that in the "Everything else" section, we want @classNames to come last, but it's still before @actions.

File: packages/components/app/styles/components/dropdown.scss (hunk 15 above)
Comment (MelSumner) [reply packages/components/app/styles/components/dropdown.scss:111]: yes, that's why we created an issue to track it, remember?

File: packages/components/app/styles/components/dropdown.scss (hunk 21 above)
Comment (MelSumner) [reply packages/components/app/styles/components/dropdown.scss:177]: can you explain more please? It makes maintenance harder to have these (`@include`, `@mixin` etc) not at the top.

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (MelSumner) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: That's factually inaccurate. Let's schedule a call to discuss.

File: packages/components/app/styles/components/dropdown.scss (hunk 14 above)
Comment (MelSumner) [reply packages/components/app/styles/components/dropdown.scss:107]: I believe our experiments have shown that we benefit from keeping min/max width in REM. Please write a separate issue or submit a separate PR if you think this should be different.

File: packages/components/app/styles/components/dropdown.scss [hunk 32]
```diff
@@ -21,31 +21,21 @@
 
 @use '../mixins/focus-ring' as *;
 
-$hds-dropdown-toggle-base-height: 36px;
-$hds-dropdown-toggle-border-radius: 5px;
-$hds-dropdown-toggle-border-width: 1px;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss]: @MelSumner I didn't suggest to remove all the variables, just the last one (was used only once).

File: packages/components/app/styles/components/dropdown.scss (hunk 15 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:110]: My understanding is that the issue is about another problem. 
The issue is about the standardization (this will involve the designers too, we will have specific design tokens, a documentation page, etc) 
What I am suggesting here it to make sure that the value we're using for the dropdown works (and using the existing Structure/Cloud UI is a good bet).

Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:110]: I checked in Cloud UI and apparently their dropdown doesn't use a z-index (not sure how that works). I tried but in our case this would not work, we need a z-index. Let's leave it as is for now, and once we work on the adoption test we can  decide what is the right value to use here.

File: packages/components/app/styles/components/dropdown.scss (hunk 21 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:176]: in all the other cases where these focus-ring mixins are used, they're used just before the other states, so it's easy to find the "focus" state in code (it's always in the block of states), and it's easy to see if that's applied and how. See for example in the Breadcrumb ([here](https://github.com/hashicorp/design-system/blob/d15ee400dbc4827a98de8a054499b358f34b2103/packages/components/app/styles/components/breadcrumb.scss#L78-L97) and [here](https://github.com/hashicorp/design-system/blob/d15ee400dbc4827a98de8a054499b358f34b2103/packages/components/app/styles/components/breadcrumb.scss#L156-L174))

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dhaulagiri) [reply packages/components/addon/components/hds/dropdown/list-item.js:95]: Found it, but I don't think we are using this plugin currently. https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/order-in-components.md

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: OK, let me rephrase it. Considered these feedbacks:
- _Jamie White - "That third option is a little more verbose, but seems the clearest to me." + "I actually have no strong opinions on this API. Y’all are thinking about this really deeply and I trust y’all to come up with a balanced design. I would be happy with any of the APIs presented."_
- _John Cowen - "of those options I would prefer 2"_
- _Matthew Irish - "[option 3] is pretty familiar since it’s pretty close to ember-basic-dropdown which is what hcp’s current dropdown was patterned after"_
and considered these are the only feedback we have received, why would you go with option 1?

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.js:95]: OK, I'll open a PR so we can align all the components to the following rule:
_tracked properties > everything else > actions, in that order. we can say that in the "Everything else" section, we want `classNames` to come last, but it's still before `@actions`

File: packages/components/app/styles/components/dropdown.scss (hunk 14 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss]: The problem was not with `rem` or `px` but with the missing `width: max-content;` declaration (we should add it)

File: packages/components/app/styles/components/dropdown.scss [hunk 33]
```diff
@@ -44,13 +44,7 @@
 
   &:focus,
   &.is-focus {
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:46]: @MelSumner you have to remove the parent declaration/selector (otherwise there are two "focus" in the generated CSS selector)

instead of this:
```
  &:focus,
  &.is-focus {
    @include hds-focus-ring-with-pseudo-element($top: -1px, $right: -1px, $bottom: -1px, $left: -1px, $radius: 5px);
  }
```
should be this:
```
  @include hds-focus-ring-with-pseudo-element($top: -1px, $right: -1px, $bottom: -1px, $left: -1px, $radius: 5px);
```

File: packages/components/app/styles/components/dropdown.scss [hunk 34]
```diff
@@ -77,6 +71,7 @@
 
 .hds-dropdown-toggle-icon__chevron {
   margin-left: 0.25rem;
+  transition: ease-in-out 0.3s;
```
Comment (didoo) [inline packages/components/app/styles/components/dropdown.scss:74]: @MelSumner this doesn't work

**Review by didoo — APPROVED** [review]
As discussed with @MelSumner we will merge this branch in `main`.
Next week I'll do the final tunings (among other things change the API/invocation of the Dropdown following this initial PR #185) using a new feature branch.

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: After discussion with @MelSumner during code review, we decided to go with the yielded components solution.

File: packages/components/app/styles/components/dropdown.scss (hunk 32 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss]: https://github.com/hashicorp/design-system/pull/202

File: packages/components/addon/components/hds/dropdown/index.hbs (hunk 10 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/index.hbs:3]: Done in #185

File: packages/components/addon/components/hds/dropdown/list-item.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (didoo) [reply packages/components/addon/components/hds/dropdown/list-item.js:95]: Opened issue https://github.com/hashicorp/design-system/issues/197

File: packages/components/app/styles/components/dropdown.scss (hunk 23 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:207]: https://github.com/hashicorp/design-system/pull/202

File: packages/components/app/styles/components/dropdown.scss (hunk 14 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss]: https://github.com/hashicorp/design-system/pull/202

File: packages/components/app/styles/components/dropdown.scss (hunk 16 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:121]: https://github.com/hashicorp/design-system/pull/202

File: packages/components/app/styles/components/dropdown.scss (hunk 27 above)
Comment (didoo) [reply packages/components/app/styles/components/dropdown.scss:231]: https://github.com/hashicorp/design-system/pull/204
